### PR TITLE
Add support of Mistral-7B into TT-Transformers 

### PR DESCRIPTION
**Pull request by pprajapatiTT**
Ticket
https://github.com/tenstorrent/tt-metal/issues/19490

Problem description
Add support of Mistral-7B into TT-Transformers and deprecates the old stale (lower-perf) implementation.

What's changed
Remove deprecated implementation of Mistral-7B (in `/models/demos/wormhole/mistral-7b/)
Add Mistral-7B support to TT-Transformers
Validated Mistral-7B-Instruct-v0.3 using TT-Transformers via HuggingFace interface
Add performance and accuracy numbers
Add CI tests
Fix the tests for HuggingFace based models

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: .github/workflows/t3000-demo-tests-impl.yaml [hunk 1]
```diff
@@ -30,6 +30,7 @@ jobs:
           { name: "t3k_falcon7b_tests", arch: wormhole_b0, cmd: run_t3000_falcon7b_tests, timeout: 90, owner_id: U05RWH3QUPM}, #Salar Hosseini
           { name: "t3k_mixtral_tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 50, owner_id: U03PUAKE719}, # Miguel Tairum
           { name: "t3k_resnet50_tests", arch: wormhole_b0, cmd: run_t3000_resnet50_tests, timeout: 50, owner_id: U0837MYG788}, # Marko Radosavljevic
+          { name: "t3k_mistral_tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 50, owner_id: U03PUAKE719}, # Miguel Tairum
```
Comment (yieldthought) [inline .github/workflows/t3000-demo-tests-impl.yaml:33]: Please change this to your slack member id:
image

File: .github/workflows/t3000-frequent-tests-impl.yaml [hunk 2]
```diff
@@ -35,6 +35,7 @@ jobs:
           # { name: "t3k llama3_70b tests", arch: wormhole_b0, cmd: run_t3000_llama3_70b_tests, timeout: 45, owner_id: U03FJB5TM5Y}, #Colman Glagovich  # FIXME issue #14934
           { name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 60, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
           { name: "t3k resnet tests", arch: wormhole_b0, cmd: run_t3000_resnet_tests, timeout: 30, owner_id: U052J2QDDKQ}, #Pavle Josipovic
+          { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 60, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
```
Comment (yieldthought) [inline .github/workflows/t3000-frequent-tests-impl.yaml:38]: Use your ID

File: .github/workflows/t3000-perplexity-tests-impl.yaml [hunk 3]
```diff
@@ -27,6 +27,7 @@ jobs:
           { name: "t3k_falcon40b_tests", arch: wormhole_b0, cmd: run_t3000_falcon40b_perplexity_tests, timeout: 300, owner_id: U053W15B6JF}, # Djordje Ivanovic
           { name: "t3k_llama_70b_tests", arch: wormhole_b0, cmd: run_t3000_llama70b_perplexity_tests, timeout: 300, owner_id: U03FJB5TM5Y}, # Colman Glagovich
           { name: "t3k_mixtral_tests", arch: wormhole_b0, cmd: run_t3000_mixtral8x7b_perplexity_tests, timeout: 300, owner_id: U03PUAKE719}, # Miguel Tairum
+          { name: "t3k_mistral_tests", arch: wormhole_b0, cmd: run_t3000_mistral_perplexity_tests, timeout: 300, owner_id: U03PUAKE719}, # Miguel Tairum
```
Comment (yieldthought) [inline .github/workflows/t3000-perplexity-tests-impl.yaml:30]: Use your ID

File: .github/workflows/t3000-unit-tests-impl.yaml [hunk 4]
```diff
@@ -35,6 +35,7 @@ jobs:
           { name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
           { name: "t3k grok tests", arch: wormhole_b0, cmd: run_t3000_grok_tests, timeout: 30, owner_id: U03HY7MK4BT}, #Mark O'Connor
           { name: "t3k unet shallow tests", arch: wormhole_b0, cmd: run_t3000_unet_shallow_tests, timeout: 30, owner_id: U06ECNVR0EN}, #Evan Smal
+          { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 30, label: pipeline-functional, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
```
Comment (yieldthought) [inline .github/workflows/t3000-unit-tests-impl.yaml:38]: Use your ID

File: models/tt_transformers/tests/test_model.py [hunk 5]
```diff
@@ -129,18 +134,25 @@ def test_model_inference(
             "llama31_8b": 0.9997,
             "llama32_11b": 0.9995,
             "llama31_70b": 0.9997,
+            "Mistral-7B": 0.68,
```
Comment (yieldthought) [inline models/tt_transformers/tests/test_model.py:137]: Please create an issue to investigate this. This does not look ok.

Comment (yieldthought) [reply models/tt_transformers/tests/test_model.py:137]: Run the accuracy test for a larger prefill length to stress the kv-cache more. If accuracy drops from the 91/98 we see at the short length, this is a good sign we have a bug that we need to fix. If it doesn't, perhaps the test is comparing the kv cache incorrectly.

File: models/tt_transformers/tests/test_model_prefill.py [hunk 6]
```diff
@@ -95,6 +88,17 @@ def test_model_inference(
     model_args = ModelArgs(mesh_device, max_batch_size=batch_size, optimizations=optimizations, max_seq_len=seq_len)
     tokenizer = model_args.tokenizer
 
+    perf_pcc_map = {"Mistral-7B-Instruct-v0.3": 0.73}
```
Comment (yieldthought) [inline models/tt_transformers/tests/test_model_prefill.py:91]: As with the kv-cache, I do not think this is an ok value. We should figure out why our output is so wildly different to the reference model.

File: models/tt_transformers/tt/common.py [hunk 7]
```diff
@@ -120,18 +120,21 @@ def preprocess_inputs_prefill(
...
     for i, encoded in enumerate(encoded_prompts):
         # Prefill size is nearest power of 2
         prefill_seq_len = max(2 ** math.ceil(math.log(len(encoded), 2)), 128)
+        max_prefill_seq_len = max(max_prefill_seq_len, prefill_seq_len)
+        prefill_lens.append(prefill_seq_len)
 
+    for i, encoded in enumerate(encoded_prompts):
         # Initial prefill tensors full of pad tokens
-        input_tokens_prefill_i = torch.full((1, prefill_seq_len), 0, dtype=torch.int32)
+        input_tokens_prefill_i = torch.full((1, max_prefill_seq_len), 0, dtype=torch.int32)
```
Comment (yieldthought) [inline models/tt_transformers/tt/common.py:132]: As in https://github.com/tenstorrent/tt-metal/pull/19435 I'm not sure this is the right thing to do. I don't think we need or want all the prefill lens to be the same as they are processed batch 1. Whatever introduced a torch.stack should be investigated and perhaps fixed there instead.

**Review by yieldthought — CHANGES_REQUESTED** [review]

**Comment by pprajapatiTT** [comment]
@mtairum , @yieldthought, @skhorasganiTT: i implemented the padding in demo (removed from common.py) see: https://github.com/tenstorrent/tt-metal/pull/19995/commits/328399b071e7607e10a8428a594b3d6985e73be8

File: .github/workflows/single-card-demo-tests-impl.yaml
Comment (mtairum) [inline .github/workflows/single-card-demo-tests-impl.yaml:1]: Remove the added line

File: models/tt_transformers/tt/model_config.py [hunk 8]
```diff
@@ -1838,7 +1878,7 @@ def reference_decoder(self):
         else:
             model = self.reference_transformer(wrap=False)
             layer = model.model.layers[0]
-            wrapper = HfDecoderWrapper(layer, self.head_dim, model.model.rotary_emb)
+            wrapper = HfDecoderWrapper(layer, self.head_dim, layer.self_attn.rotary_emb)
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py:1881]: This needs to be generalized as a wider TT-Transformers refactor for HF. Other models are called model.model. and I suspect other models will have something else 😢 

Can you change this to be `layer.self_atttn` if the model is mistral and keep it the older way otherwise?

Also add a comment with a TODO: `...  # TODO Generalize for other HF models`

File: models/tt_transformers/tt/model_config.py [hunk 9]
```diff
@@ -1997,7 +2037,7 @@ def __init__(self, decoder, head_dim, rotary_emb):
 
     def forward(self, x, start_pos, freqs_cis_i, mask=None):
         position_ids = torch.tensor([list(range(start_pos, start_pos + x.shape[1]))] * x.shape[0])
-        position_embeddings = self.rotary_emb(x, position_ids)
+        position_embeddings = self.rotary_emb(x, x.shape[1])
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py:2040]: Is this doing the rotary embeding with the full input positions len, and then we run decoder just with the position ids?

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
@pprajapatiTT Review finished. Please address the 2 small comments I left.

Re-running the relevant pipelines here:
- [x] [All-post-commit](https://github.com/tenstorrent/tt-metal/actions/runs/14445239324/job/40508912390)
- [ ] [T3K](https://github.com/tenstorrent/tt-metal/actions/runs/14453143174)
- [ ] [Single-card](https://github.com/tenstorrent/tt-metal/actions/runs/14453163217) <- Only added test to single-demo pipeline. But running all pipelines to make sure LLama3 is not affected, since this toches TT-Transformers.

File: tests/scripts/t3000/run_t3000_perplexity_tests.sh [hunk 10]
```diff
@@ -116,6 +116,27 @@ run_t3000_llama3_perplexity_tests_single_card() {
...
+  # This one runs all the T3K tests
+  fail=0
+  start_time=$(date +%s)
+
+  echo "LOG_METAL: Running run_t3000_mistral_perplexity_tests"
+
+  wh_arch_yaml=wormhole_b0_80_arch_eth_dispatch.yaml
+  tt_cache_path="/mnt/MLPerf/tt_dnn-models/Mistral/TT_CACHE/Mistral-7B-Instruct-v0.3"
+  hf_model="/mnt/MLPerf/tt_dnn-models/Mistral/hub/models--mistralai--Mistral-7B-Instruct-v0.3/snapshots/e0bc86c23ce5aae1db576c8cca6f06f1f73af2db"
+  WH_ARCH_YAML=$wh_arch_yaml TT_CACHE_PATH=$tt_cache_path HF_MODEL=$hf_model pytest -n auto models/tt_transformers/tests/test_accuracy.py --timeout=3600 ; fail+=$?
```
Comment (tt-rkim) [inline tests/scripts/t3000/run_t3000_perplexity_tests.sh:129]: As long as there's one test here, we can keep logging by getting rid of `-n auto`, although not sure if the timeouts kick in

File: .github/workflows/t3000-unit-tests-impl.yaml [hunk 11]
```diff
@@ -35,6 +35,7 @@ jobs:
           { name: "t3k mixtral tests", arch: wormhole_b0, cmd: run_t3000_mixtral_tests, timeout: 30, owner_id: U03PUAKE719}, #Miguel Tairum Cruz
           { name: "t3k grok tests", arch: wormhole_b0, cmd: run_t3000_grok_tests, timeout: 30, owner_id: U03HY7MK4BT}, #Mark O'Connor
           { name: "t3k unet shallow tests", arch: wormhole_b0, cmd: run_t3000_unet_shallow_tests, timeout: 30, owner_id: U06ECNVR0EN}, #Evan Smal
+          { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 30, label: pipeline-functional, owner_id: U0896VBAKFC}, # Pratikkumar Prajapati
```
Comment (williamlyTT) [inline .github/workflows/t3000-unit-tests-impl.yaml:38]: Small nit: no need for `label: pipeline-functional`, already included by default.
```suggestion
          { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 30, owner_id: U0896VBAKFC}, # Pratikkumar Prajapati
```

**Comment by pprajapatiTT** [comment]
I started these CI pipelines after my changes to address the comments.

T3k: https://github.com/tenstorrent/tt-metal/actions/runs/14459339135
Single-card: https://github.com/tenstorrent/tt-metal/actions/runs/14459348279
All post-commit-tests: https://github.com/tenstorrent/tt-metal/actions/runs/14459354502

File: .github/workflows/fast-dispatch-full-regressions-and-models-impl.yaml [hunk 12]
```diff
@@ -124,15 +124,12 @@ jobs:
             -e ARCH_NAME=wormhole_b0
             -e LD_LIBRARY_PATH=${{ github.workspace }}/build/lib
             -e GTEST_OUTPUT=xml:generated/test_reports/
-          # Llama3 has a single pytest for multiple llama models, hence it requires calling it multiple times.
-          # Due to host OOM issues in CI vm, we currently only run llama-1B in the model matrix.
+            -e WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml
```
Comment (mtairum) [inline .github/workflows/fast-dispatch-full-regressions-and-models-impl.yaml:127]: Just a FYI @williamlyTT @tt-rkim. 

In main (before this PR) all models after mistral were already using `WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml` since we were exporting it. If we do not export this most models will fail 🥲 
I think most model owners were oblivious to this, since it did seem specific to mistral.

**Comment by pprajapatiTT** [comment]
CIs:
Post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/14603850758
T3k: https://github.com/tenstorrent/tt-metal/actions/runs/14603868289
Single-card: https://github.com/tenstorrent/tt-metal/actions/runs/14603881571

File: models/tt_transformers/tests/test_model.py [hunk 13]
```diff
@@ -132,18 +138,25 @@ def test_model_inference(
...
```
Comment (yieldthought) [inline models/tt_transformers/tests/test_model.py:141]: 0.68 is terrible PCC. Something is wrong here, maybe in the test/reference PCC calculations as the model accuracy looks otherwise good. If the test is wrong I'd rather fix it or even skip it than have this as the target PCC.

**Review by yieldthought — CHANGES_REQUESTED** [review]

File: tests/scripts/single_card/run_single_card_demo_tests.sh [hunk 14]
```diff
@@ -9,7 +9,9 @@ run_falcon7b_func() {
 
 run_mistral7b_func() {
 
-  WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest models/demos/wormhole/mistral7b/demo/demo.py
+  mistral7b=/mnt/MLPerf/tt_dnn-models/Mistral/hub/models--mistralai--Mistral-7B-Instruct-v0.3/snapshots/e0bc86c23ce5aae1db576c8cca6f06f1f73af2db
```
Comment (tt-rkim) [inline tests/scripts/single_card/run_single_card_demo_tests.sh:12]: I don't understand. We're removing mistral7b functionality tests from single card demo tests?

File: tests/scripts/t3000/run_t3000_demo_tests.sh [hunk 15]
```diff
@@ -122,6 +122,27 @@ run_t3000_falcon7b_tests(){
   fi
 }
 
+run_t3000_mistral_tests() {
+  # Record the start time
+  fail=0
+  start_time=$(date +%s)
+
+  echo "LOG_METAL: Running run_t3000_mistral_demo_tests"
```
Comment (tt-rkim) [inline tests/scripts/t3000/run_t3000_demo_tests.sh:130]: No need for all this fluff around the function now since no one actually uses `run_t3000_tests`. Please just define the variables and call the pytest command.

File: models/tt_transformers/tests/test_model.py (hunk 13 above)
Comment (yieldthought) [reply models/tt_transformers/tests/test_model.py:151]: No this is correctly calling model.reference_transformer() which loads the huggingface reference model to compare against. I don't understand why the PCC is so bad here. I'd think rather something about converting the kv cache values from the huggingface reference back and comparing them to ours. @pprajapatiTT Our QKV matrix is reordered because our RoPE implementation follows Meta's and has the real and imaginary pairs interleaved, wherease huggingface puts all the imaginary ones in the second half. I think you'd need to load_checkpoints.reverse_permute the HF KV caches before comparing them to ours, for example.

**Review by yieldthought — CHANGES_REQUESTED** [review]
I'm increasingly convinced the bad PCC values are just demonstrating a bug in this code that we should fix.

**Review by tt-rkim — APPROVED** [review]

**Review by yieldthought — APPROVED** [review]

**Review by patrickroberts — APPROVED** [review]
